src/htm/encoders/DateEncoder.cpp

```diff
@@ -0,0 +1,420 @@
+    Real64 customDay = 0.0;
+    if (customDays_.find(timeinfo.tm_wday) != customDays_.end()) {
+        customDay = 1.0;
+    }
+    customDay_output = SDR(customDaysEncoder_->dimensions);
+    customDaysEncoder_->encode(customDay, customDay_output);
+    VERBOSE << "  custom Day: " << customDay << " ==> " << customDay_output;
+    sdrs.push_back(&customDay_output);
+  }
+
+  if (holidayEncoder_) {
+    // A "continuous" binary value. = 1 on the holiday itself and smooth ramp
+    //  0->1 on the day before the holiday and 1->0 on the day after the holiday.
+    // holidays is a list of holidays that occur on a fixed date every year
+    Real64 val = 0.0;
+    double SECONDS_PER_DAY = 86400.0;
+    time_t input = std::mktime(&timeinfo);
+    for (auto h : args_.holiday_dates) {
+      std::time_t hdate;
+      struct tm holidayinfo;
+      memset(&holidayinfo, 0, sizeof(holidayinfo));
+      if (h.size() == 3) {
+        holidayinfo.tm_year = h[0] - 1900;
+        holidayinfo.tm_mon = h[1] - 1;
+        holidayinfo.tm_mday = h[2];
+        holidayinfo.tm_isdst = -1;
+        hdate = mktime(&holidayinfo);
+      } else {
+        holidayinfo.tm_year = timeinfo.tm_year;
+        holidayinfo.tm_mon = h[0] - 1;
+        holidayinfo.tm_mday = h[1];
+        holidayinfo.tm_isdst = -1;
+        hdate = mktime(&holidayinfo);
+      }
+      if (input > hdate) {
+        // start of holiday is in the past.
+        std::time_t diff = input - hdate;
+        if (diff < SECONDS_PER_DAY) {
+          // return 1 on the holiday itself
+          val = 1.0;
+          break;
+        } else if (diff < SECONDS_PER_DAY * 2.0) {
+          // Next day, ramp smoothly from 1 -> 0
+          val = 1.0 + ((diff - SECONDS_PER_DAY) / SECONDS_PER_DAY);
+          break;
+        }
+      } else {
+        // start of holiday is in the future.
+        std::time_t diff = hdate - input;
+        if (diff < SECONDS_PER_DAY) {
+          // holiday starts tomarrow
+          // ramp smoothly from 0 -> 1 on the previous day
+          val = 1.0 - diff / SECONDS_PER_DAY;
+          break;
+        }
+      }
+    }
+    holiday_output = SDR(holidayEncoder_->dimensions);
+    holidayEncoder_->encode(val, holiday_output);
+    VERBOSE << "  holiday: " << val << " ==> " << holiday_output;
+    sdrs.push_back(&holiday_output);
+  }
+  if (timeOfDayEncoder_) {
+    Real64 timeOfDay = timeinfo.tm_hour + timeinfo.tm_min / 60.0f + timeinfo.tm_sec / (60.0 * 60.0);
+    timeOfDay_output = SDR(timeOfDayEncoder_->dimensions);
+    timeOfDayEncoder_->encode(timeOfDay, timeOfDay_output);
+    VERBOSE << "  timeOfDay: " << timeOfDay << "hrs ==> " << timeOfDay_output;
+    sdrs.push_back(&timeOfDay_output);
+  }
+
+  output = concatenateSDR(sdrs);
+}
+
+// Copy all of the component SDR's into one concatinated SDR.
+// The component SDR's have different dimensions so the output is a flat SDR.
+SDR DateEncoder::concatenateSDR(const std::vector<SDR *> &sdrs) const {
```

Comment:
IIRC: SDR's have a method for doing this.

Reply:
Yes, I wanted to used the method in the SDR but it required that all parts have the same width.  All of these are different.

Reply:
hmmm, I went back and looked at the SDR again.  I guess along the axis they do not need to be the same.  So maybe I could use it.   Let me give it a try.

Reply:
I'm glad to hear you were able to get the SDR.concatenate method to work!  I was worried that you found a flaw in the API of the SDR.concatenate method which prevented it from being used.

> it required that all parts have the same width. All of these are different.

If the dimensions of the SDR's are truly incompatible then the user will need to reshape all of their SDR's to 1-dimension first.  I think there is a method to do that called `SDR.flatten()`.